app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -102,258 +101,296 @@ public DataPullTask(String username, String password,
+                    "User Sync failed due to bad payload|" + e.getMessage());
+            return new ResultAndError<>(PullTaskResult.BAD_DATA_REQUIRES_INTERVENTION,
+                    e.getLocalizedMessage());
+        } catch (InvalidStructureException e) {
+            e.printStackTrace();
+            Logger.log(AndroidLogger.TYPE_USER,
+                    "User Sync failed due to bad payload|" + e.getMessage());
+            return new ResultAndError<>(PullTaskResult.BAD_DATA, e.getMessage());
+        } catch (UnfullfilledRequirementsException e) {
+            e.printStackTrace();
+            Logger.log(AndroidLogger.TYPE_ERROR_ASSERTION,
+                    "User sync failed oddly, unfulfilled reqs |" + e.getMessage());
+            return null;
+        } catch (IllegalStateException e) {
+            e.printStackTrace();
+            Logger.log(AndroidLogger.TYPE_ERROR_ASSERTION,
+                    "User sync failed oddly, ISE |" + e.getMessage());
+            return null;
+        } catch (RecordTooLargeException e) {
+            e.printStackTrace();
+            Logger.log(AndroidLogger.TYPE_ERROR_ASSERTION,
+                    "Storage Full during user sync |" + e.getMessage());
+            return new ResultAndError<>(PullTaskResult.STORAGE_FULL);
+        }
+    }
+
+    private void handleLoginNeededOnSuccess() {
+        if (loginNeeded) {
+            // This is currently necessary to make sure that data is encoded, but there is
+            // probably a better way to do it
+            CommCareApplication._().startUserSession(
+                    ByteEncrypter.unwrapByteArrayWithString(ukrForLogin.getEncryptedKey(), password),
+                    ukrForLogin, false);
+            wasKeyLoggedIn = true;
+        }
+    }
+
+    /**
+     * @return the proper result, or null if we have not yet been able to determine the result to
+     * return
+     * @throws IOException
+     */
+    private ResultAndError<PullTaskResult> handleBadLocalState(AndroidTransactionParserFactory factory) {
+        Pair<Integer, String> returnCodeAndMessageFromRecovery = recover(requestor, factory);
+        int returnCode = returnCodeAndMessageFromRecovery.first;
+        String failureReason = returnCodeAndMessageFromRecovery.second;
+
+        if (returnCode == PROGRESS_DONE) {
+            // Recovery was successful
+            onSuccessfulSync();
+            return new ResultAndError<>(PullTaskResult.DOWNLOAD_SUCCESS);
+        } else if (returnCode == PROGRESS_RECOVERY_FAIL_SAFE || returnCode == PROGRESS_RECOVERY_FAIL_BAD) {
+            wipeLoginIfItOccurred();
             this.publishProgress(PROGRESS_DONE);
-            return new ResultAndError<>(responseError);
-        } finally {
-            CommCareSessionService.sessionAliveLock.unlock();
+            return new ResultAndError<>(PullTaskResult.UNKNOWN_FAILURE, failureReason);
+        } else {
+            wipeLoginIfItOccurred();
+            return null;

```

Comment:
Same reasoning as above. I can explain in person if I'm not doing a good job of it here.